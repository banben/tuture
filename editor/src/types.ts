export type ExplainType = 'pre' | 'post';

export type EditMode = 'edit' | 'preview';
