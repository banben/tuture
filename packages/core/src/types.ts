import { Element } from 'tuture-slate';

import { DiffFile } from './diff';

export interface RawDiff {
  commit: string;
  diff: DiffFile[];
}

export interface Explain extends Element {
  type: 'explain';
  fixed: true;
}

export interface DiffBlock extends Element {
  type: 'diff-block';
  file: string;
  commit: string;
  hiddenLines?: [number, number][];
}

export interface File extends Element {
  type: 'file';
  file: string;
  display?: boolean;
  children: [Explain, DiffBlock, Explain];
}

export interface Meta {
  name: string;
  description?: string;
  id: string;
  cover?: string;
  created?: Date;
  topics?: string[];
  categories?: string[];
  github?: string;
}

export interface Article extends Meta {}

export interface StepTitle extends Element {
  type: 'heading-two';
  commit: string;
  id: string;
  fixed: true;
}

export type StepChild = StepTitle | Explain | File;

export interface Step extends Element {
  type: 'step';
  id: string;
  articleId?: string | null;
  outdated?: boolean;
  commit: string;
  children: StepChild[];
}

export interface Remote {
  name: string;
  refs: {
    fetch: string;
    push: string;
  };
}

export interface Collection extends Meta {
  version?: string;
  articles: Article[];
  remotes?: Remote[];
  steps: Step[];
}

export interface TutureConfig {
  ignoredFiles: string[];
  port: number;
}
