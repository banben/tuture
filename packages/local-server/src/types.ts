export interface TocArticleItem {
  type: 'article';
  id: string;
  name: string;
  level: number;
}

export interface TocStepItem {
  type: 'step';
  id: string;
  name: string;
  level: number;
  number: number;
  articleId?: string | null;
  outdated?: boolean;
}

export type TocItem = TocArticleItem | TocStepItem;

export interface CollectionStep {
  key: string;
  id: string;
  title: string;
  articleId?: string | null;
  articleIndex: number;
  articleName: string;
}
