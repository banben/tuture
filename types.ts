export interface Explain {
  pre?: string;
  post?: string;
}

export interface Diff {
  file: string;
  explain?: Explain;
  display?: boolean;
}

export interface TutureMeta {
  name: string;
  topics?: string[];
  categories?: string[];
  description?: string;
  created?: Date;
  updated?: Date;
  cover?: string;
  github?: string;
}

export interface Split extends TutureMeta {
  start: string;
  end: string;
}

export interface Commit {
  name: string;
  commit: string;
}

export interface Step extends Commit {
  explain?: Explain;
  diff: Diff[];
  outdated?: boolean;
}

export interface Tuture extends TutureMeta {
  id?: string;
  splits?: Split[];
  steps: Step[];
}

export interface TutureConfig {
  ignoredFiles: string[];
  port: number;
}
