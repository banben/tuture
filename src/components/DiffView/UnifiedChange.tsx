import React, { PureComponent } from 'react';
import classnames from 'classnames';

import { Change as ChangeType } from '../ContentItem';
import { Cell } from './SplitChange';

import {
  computeOldLineNumber,
  computeNewLineNumber,
} from './utils';

interface UnifiedChangeProps {
  change: ChangeType;
}

export default class UnifiedChange extends PureComponent<UnifiedChangeProps> {
  renderGutterCell = ({
    hide,
    className,
    lineNumber,
    gutterAnchor,
    anchorID,
    ...props
  }: Cell): React.ReactNode => {
    if (hide) {
      return null;
    }

    return (
      <td className={className} data-line-number={lineNumber} {...props}>
        {
          gutterAnchor
          ? <a href={`#${anchorID}`} data-line-number={lineNumber} />
          : null
        }
      </td>
    );
  }

  render() {
    const {
      change,
    } = this.props;

    const { type, content } = change;
    const oldLine = computeOldLineNumber(change);
    const oldLineNumber = oldLine === -1 ? undefined : oldLine;
    const newLine = computeNewLineNumber(change);
    const newLineNumber = newLine === -1 ? undefined : newLine;

    const gutterClassName = classnames(
      'diff-gutter',
      `diff-gutter-${type}`,
    );
    const codeClassName = classnames(
      'diff-code',
      `diff-code-${type}`,
    );

    return (
      <tr className={classnames('diff-line')}>
        {
          this.renderGutterCell({
            className: gutterClassName,
            lineNumber: oldLineNumber,
          })
        }
        {
          this.renderGutterCell({
            className: gutterClassName,
            lineNumber: newLineNumber
          })
        }
        <td className={codeClassName}>{content}</td>
      </tr>
    );
  }
}
