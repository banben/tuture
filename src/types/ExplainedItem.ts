type ExplainType = 'pre' | 'post';
type EditMode = 'edit' | 'preview';

type ToolType =
  | 'b'
  | 'i'
  | 'h'
  | 'list'
  | 'numbered list'
  | 'blockquotes'
  | 'link'
  | 'img'
  | 'code'
  | 'block code';

export { ExplainType, EditMode, ToolType };
