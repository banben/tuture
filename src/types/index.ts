interface ChangedFile {
  file: string;
  explain: string;
  collapse?: boolean;
}

interface TutureMeta {
  name: string;
  language: string;
  version: string;
  description: string;
  maintainer: string;
  topics: string[];
}

interface Commit {
  name: string;
  commit: string;
}

interface Step extends Commit {
  explain: string;
  diff: ChangedFile[];
}

interface Tuture extends TutureMeta {
  steps: Step[];
}

interface Change {
  content: string;
  type: string;
  isInsert?: boolean;
  isNormal?: boolean;
  isDelete?: boolean;
  lineNumber?: number;
  oldLineNumber?: number;
  newLineNumber?: number;
}

interface Hunk {
  content: string;
  newLines: number;
  oldLines: number;
  oldStart: number;
  newStart: number;
  changes: Change[];
  isPlain?: boolean;
}

export { ChangedFile, TutureMeta, Commit, Step, Tuture, Change, Hunk };
