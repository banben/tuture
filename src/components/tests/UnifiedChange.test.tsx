import React from 'react';
import renderer from 'react-test-renderer';

import UnifiedChange from '../DiffView/UnifiedChange';
import { Change } from '../../types/index';
import { shallow } from 'enzyme';

describe('Steps component/dom/functionality/events level behave normally', () => {
  const change: Change = {
    type: 'insert',
    content: 'I am the coolest one',
    isInsert: true,
    lineNumber: 0,
  };

  describe('component-level behave normally', () => {
    test('UnifiedChange render successfully', () => {
      const component = renderer.create(
        <UnifiedChange
          change={change}
        />
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('dom-level behave normally', () => {
    const wrapper = shallow(
      <UnifiedChange
        change={change}
      />
    );

    test('should render realted dom node', () => {
      expect(wrapper.find('.diff-line')).toHaveLength(1);
      expect(wrapper.find('.diff-code-insert')).toHaveLength(1);
      expect(wrapper.find('.diff-gutter-insert')).toHaveLength(2);
    });
  });
});
