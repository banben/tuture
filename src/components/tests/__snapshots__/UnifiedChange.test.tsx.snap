// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps behave normally on component/dom/functionality/events level component-level behave normally UnifiedChange render successfully 1`] = `
<tr
  className="diff-line"
>
  <td
    className="diff-gutter diff-gutter-insert"
  />
  <td
    className="diff-gutter diff-gutter-insert"
    data-line-number={0}
  />
  <td
    className="diff-code diff-code-insert"
  >
    I am the coolest one
  </td>
</tr>
`;
